src/expressions/operators/arithmetic/BinaryEvaluationFunctionMap.ts

```diff
@@ -0,0 +1,238 @@
+import { ValueType } from '../../../expressions/dataTypes/Value';
+import {
+	addDuration as addDurationToDateTime,
+	subtract as dateTimeSubtract,
+	subtractDuration as subtractDurationFromDateTime,
+} from '../../../expressions/dataTypes/valueTypes/DateTime';
+import {
+	add as dayTimeDurationAdd,
+	divide as dayTimeDurationDivide,
+	divideByDayTimeDuration as dayTimeDurationDivideByDayTimeDuration,
+	multiply as dayTimeDurationMultiply,
+	subtract as dayTimeDurationSubtract,
+} from '../../../expressions/dataTypes/valueTypes/DayTimeDuration';
+import {
+	add as yearMonthDurationAdd,
+	divide as yearMonthDurationDivide,
+	divideByYearMonthDuration as yearMonthDurationDivideByYearMonthDuration,
+	multiply as yearMonthDurationMultiply,
+	subtract as yearMonthDurationSubtract,
+} from '../../../expressions/dataTypes/valueTypes/YearMonthDuration';
+
+/**
+ * A hash function that is used to create the keys for the ruleMap.
+ * @param left the ValueType of the left part of the operator
+ * @param right the ValueType of the right part of the operator
+ * @param op the operator
+ * @returns a number that can be used as a key
+ */
+export function hash(left: ValueType, right: ValueType, op: string): number {
+	return (
+		((left as number) << 20) +
```

Comment:
Maybe consider making these constants?

Reply:
I'm not sure I understand the review. Do you mean to change the 20, 12, and 8 to constants?

Reply:
Ah, sorry, yeah, something like `const AMOUNT_OF_TYPES` or something, so that someone can deduce what the 20 means here. And why they should change it. If we start implementing XPath 4, we might get a bunch of new types, so we might have to change these constants in order to have a good hasing constant again

Reply:
Ooh yeah that makes sense, we'll do that

Reply:
I believe they do not really have meanings beyond just being some integer that simply shifts the data over a bit to make sure they do not overlap (or at least reduce it as much as possible)
Now that I look at it, they probably should also be xor-ed together, that than added.

Reply:
If they do not have meaning, remove them. I do not like meaningless code. Or set them all to 5, my fav number. These are magic numbers, which I dislike.

Jokes aside, just name them....